### PR TITLE
Fix CompareTo/Equals when dealing with Empty Span or Span wrapping a null string

### PR DESCRIPTION
**Pull request by ahsonkhan**
- Added missing Debug.Asserts to catch when an empty span is being passed to Interop code (which could result in NullReferenceException).
- Fixed Span.CompareTo and Span.Equals by adding checks for `length == 0` and addressing them ahead of time. Resolved based on the discussion in https://github.com/dotnet/corefx/issues/27350.
- Fixed the implementation of Trim, TrimStart, TrimEnd so that its behaviour is identical to the string APIs when trimChars is null/empty.

From https://msdn.microsoft.com/en-us/library/d4tt83f9(v=vs.110).aspx
> The string that remains after all occurrences of the characters in the trimChars parameter are removed from the start and end of the current string. If trimChars is null or an empty array, white-space characters are removed instead.

Added tests in corefx: https://github.com/dotnet/corefx/pull/28347

cc @KrzysztofCwalina, @jkotas, @tarekgh, @stephentoub, @joshfree, @VSadov

### CODE REVIEW

File: src/mscorlib/shared/System/Globalization/CompareInfo.Unix.cs [hunk 1]
```diff
@@ -181,6 +181,8 @@ internal static unsafe int LastIndexOfOrdinalCore(string source, string value, i
         private static unsafe int CompareStringOrdinalIgnoreCase(char* string1, int count1, char* string2, int count2)
         {
             Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert(string1 != null);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.Unix.cs]: The assert in this file are unnecessary. The Unix globalizations API handle NULL with zero-sized buffer just fine.

File: src/mscorlib/shared/System/Globalization/CompareInfo.cs [hunk 2]
```diff
@@ -393,13 +393,33 @@ internal int Compare(ReadOnlySpan<char> string1, string string2, CompareOptions
 
         internal virtual int CompareOptionNone(ReadOnlySpan<char> string1, ReadOnlySpan<char> string2)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.cs]: `virtual` is not necessary here. You can remove it.

File: src/mscorlib/shared/System/Globalization/CompareInfo.cs [hunk 3]
```diff
@@ -393,13 +393,33 @@ internal int Compare(ReadOnlySpan<char> string1, string string2, CompareOptions
...
+            }
+            if (string2.IsEmpty)
+                return 1;
+
             return _invariantMode ?
                 string.CompareOrdinal(string1, string2) :
                 CompareString(string1, string2, CompareOptions.None);
         }
 
         internal virtual int CompareOptionIgnoreCase(ReadOnlySpan<char> string1, ReadOnlySpan<char> string2)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.cs]: Same here. Virtual is unnecessary.

File: src/mscorlib/shared/System/Globalization/CompareInfo.cs [hunk 4]
```diff
@@ -895,12 +917,16 @@ public unsafe virtual int IndexOf(string source, string value, int startIndex, i
         internal virtual int IndexOfOrdinal(ReadOnlySpan<char> source, ReadOnlySpan<char> value, bool ignoreCase)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.cs]: Virtual is unnecessary.

**Comment by jkotas** [comment]
> If trimChars is null or an empty array, white-space characters are removed instead.

I think it is pretty questionable behavior. I would consider it a design bug. I am not sure whether we want to be replicating it...

File: src/mscorlib/shared/System/Globalization/CompareInfo.cs [hunk 5]
```diff
@@ -393,13 +393,33 @@ internal int Compare(ReadOnlySpan<char> string1, string string2, CompareOptions
 
         internal virtual int CompareOptionNone(ReadOnlySpan<char> string1, ReadOnlySpan<char> string2)
         {
+            // Check for empty span or span from a null string
+            if (string1.IsEmpty)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.cs]: You can just do this as:

```
                    if (span.Length == 0 || string2.Length == 0)
                        return string1.Length - string2.Length;
```

It is how the same check is done in MemoryExtensions.Fast.cs.

File: src/mscorlib/shared/System/Globalization/CompareInfo.cs [hunk 6]
```diff
@@ -895,12 +917,16 @@ public unsafe virtual int IndexOf(string source, string value, int startIndex, i
         internal virtual int IndexOfOrdinal(ReadOnlySpan<char> source, ReadOnlySpan<char> value, bool ignoreCase)
         {
             Debug.Assert(!_invariantMode);
+            Debug.Assert(!source.IsEmpty);
+            Debug.Assert(!value.IsEmpty);
             return IndexOfOrdinalCore(source, value, ignoreCase);
         }
 
         internal unsafe virtual int IndexOf(ReadOnlySpan<char> source, ReadOnlySpan<char> value, CompareOptions options)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Globalization/CompareInfo.cs]: virtual is unnecessary.

File: src/mscorlib/shared/System/Globalization/CompareInfo.Unix.cs (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Globalization/CompareInfo.Unix.cs]: (You can keep them if you would like though.)

**Comment by ahsonkhan** [comment]
@dotnet-bot test Windows_NT arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_baseline
@dotnet-bot test Windows_NT x86 Checked corefx_baseline

**Comment by ahsonkhan** [comment]
@dotnet-bot test Windows_NT arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_baseline
@dotnet-bot test Windows_NT x86 Checked corefx_baseline

**Comment by ahsonkhan** [comment]
@dotnet-bot test Windows_NT arm Cross Checked corefx_baseline Build and Test
@dotnet-bot test Windows_NT x64_arm64_altjit Checked corefx_baseline
@dotnet-bot test Ubuntu x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x86_arm_altjit Checked corefx_baseline
@dotnet-bot test Windows_NT x86 Checked corefx_baseline

**Comment by ahsonkhan** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

**Pull request merged**